### ISSUE TITLE
Write \hypertarget for labels to allow external linking to a destination within a PDF

### ISSUE THREAD

**Issue by jessetan** [sphinx-doc/sphinx#6276]
PDFs have the concept of "named targets" that can be used by an external application to jump directly to a part of the document. This is very useful to link to a specific part of a long document.

Let's say a document doc.pdf has a section with a named target "my-target", then PDF viewers can jump directly to that section by opening the url `http://example.com/doc.pdf#nameddest=my-target`. Some PDF viewers also support the (non-standard) short form `http://example.com/doc.pdf#my-target`.
This works in Firefox (with pdf.js), Chrome, Sumatra, Evince, Adobe Acrobat Reader and probably others.

It is possible to create a named destination in LaTeX by using the `\hypertarget` command from the `hyperref` package ([example on Stack Overflow](https://tex.stackexchange.com/questions/213860/how-to-generate-a-named-destination-in-pdf#answer-213862))

This could be supported by Sphinx if it would generate a `\hypertarget` around a `\label`.

Example input:
```rst
.. _this-is-label-two:

Heading 2
=========
```

Current output:
```latex
\chapter{Heading 2}
\label{\detokenize{index:heading-2}}\label{\detokenize{index:this-is-label-two}}
```

Proposed output:
```latex
\chapter{Heading 2}
\label{\detokenize{index:heading-2}}\hypertarget{this-is-label-two}{\label{\detokenize{index:this-is-label-two}}}
```

A PDF viewer can now jump directly to this section by opening http://example.com/doc.pdf#nameddest=this-is-label-two.

Note that I've removed the `index:` prefix in the hypertarget label. An external document can know the label to link to, but it is unlikely to know the original rst doc name (`index` in this case).
With this feature it would even be possible to have Intersphinx support linking between PDF files, rather than always pointing to HTML.

P.S. On my TexLive 2016, this works fine with pdflatex, but not with XeLaTeX. This is easily fixed by adding `\special{dvipdfmx:config C 0x0010}` to the preamble. See bug reports for [Xetex](https://sourceforge.net/p/xetex/bugs/132/) and [MikTex](https://sourceforge.net/p/miktex/bugs/2529/).

**Comment by tk0miya** [comment]
+1; good idea.

@jfbu what do you think?

**Comment by jfbu** [comment]
+1 too. But notice that for sectioning units this exists already, try for example [this link](https://buildmedia.readthedocs.org/media/pdf/sphinx/master/sphinx.pdf#nameddest=chapter.5)

```
https://buildmedia.readthedocs.org/media/pdf/sphinx/master/sphinx.pdf#nameddest=chapter.5
```

It works for me with Firefox and its pdf viewer. The labels are generated by hyperref package for the bookmarks. One can see them in file `foo.out` in Sphinx latex build repertory corresponding to file `foo.tex`.

The titles themselves are in UTF-16BE format, PDF-escaped ([direct reference to relevant section in hyperref pdf doc](http://mirrors.ctan.org/macros/latex/contrib/hyperref/doc/manual.pdf#nameddest=subsection.5.18)).

- advantage of such preexisting targets is that the hypertarget opens the PDF right above the section title not immediately after it,

- disadvantage is that the actual names of destinations are the ones generated by hyperref itself although they seem quite guessable as the example above which is a link using ``manual.pdf#nameddest=subsection.5.18`` shows.

- of course a user chosen target name will remain identical throughout evolution of the document.

When the `bookmark` package  is loaded after `hyperref` the used file has extension `out.ps`, and its format is a bit different but for sectioning titles it seems the same.

My remarks only explain that for sectioning units, the functionality is potentially already existing with no change at Sphinx side... if for some user added labels we want to use `\hypertarget` then we have to investigate how to position this target above caption, but I think there are already LaTeX macros for "raised links" available from `hyperref/hypcap` packages, already loaded by Sphinx. I will have to give a look.

**Comment by jfbu** [comment]
+1 too. But notice that for sectioning units this exists already, try for example [this link](https://buildmedia.readthedocs.org/media/pdf/sphinx/master/sphinx.pdf#nameddest=chapter.5)

```
https://buildmedia.readthedocs.org/media/pdf/sphinx/master/sphinx.pdf#nameddest=chapter.5
```

It works for me with Firefox and its pdf viewer. The labels are generated by hyperref package for the bookmarks. One can see them in file `foo.out` in Sphinx latex build repertory corresponding to file `foo.tex`.

The titles themselves are in UTF-16BE format, PDF-escaped ([direct reference to relevant section in hyperref pdf doc](http://mirrors.ctan.org/macros/latex/contrib/hyperref/doc/manual.pdf#nameddest=subsection.5.18)).

- advantage of such preexisting targets is that the hypertarget opens the PDF right above the section title not immediately after it,

- disadvantage is that the actual names of destinations are the ones generated by hyperref itself although they seem quite guessable as the example above which is a link using ``manual.pdf#nameddest=subsection.5.18`` shows.

- of course a user chosen target name will remain identical throughout evolution of the document.

When the `bookmark` package  is loaded after `hyperref` the used file has extension `out.ps`, and its format is a bit different but for sectioning titles it seems the same.

My remarks only explain that for sectioning units, the functionality is potentially already existing with no change at Sphinx side... if for some user added labels we want to use `\hypertarget` then we have to investigate how to position this target above caption, but I think there are already LaTeX macros for "raised links" available from `hyperref/hypcap` packages, already loaded by Sphinx. I will have to give a look.

**Comment by jfbu** [comment]
+1 too. But notice that for sectioning units this exists already, try for example [this link](https://buildmedia.readthedocs.org/media/pdf/sphinx/master/sphinx.pdf#nameddest=chapter.5)

```
https://buildmedia.readthedocs.org/media/pdf/sphinx/master/sphinx.pdf#nameddest=chapter.5
```

It works for me with Firefox and its pdf viewer. The labels are generated by hyperref package for the bookmarks. One can see them in file `foo.out` in Sphinx latex build repertory corresponding to file `foo.tex`.

The titles themselves are in UTF-16BE format, PDF-escaped ([direct reference to relevant section in hyperref pdf doc](http://mirrors.ctan.org/macros/latex/contrib/hyperref/doc/manual.pdf#nameddest=subsection.5.18)).

- advantage of such preexisting targets is that the hypertarget opens the PDF right above the section title not immediately after it,

- disadvantage is that the actual names of destinations are the ones generated by hyperref itself although they seem quite guessable as the example above which is a link using ``manual.pdf#nameddest=subsection.5.18`` shows.

- of course a user chosen target name will remain identical throughout evolution of the document.

When the `bookmark` package  is loaded after `hyperref` the used file has extension `out.ps`, and its format is a bit different but for sectioning titles it seems the same.

My remarks only explain that for sectioning units, the functionality is potentially already existing with no change at Sphinx side... if for some user added labels we want to use `\hypertarget` then we have to investigate how to position this target above caption, but I think there are already LaTeX macros for "raised links" available from `hyperref/hypcap` packages, already loaded by Sphinx. I will have to give a look.

**Comment by jfbu** [comment]
+1 too. But notice that for sectioning units this exists already, try for example [this link](https://buildmedia.readthedocs.org/media/pdf/sphinx/master/sphinx.pdf#nameddest=chapter.5)

```
https://buildmedia.readthedocs.org/media/pdf/sphinx/master/sphinx.pdf#nameddest=chapter.5
```

It works for me with Firefox and its pdf viewer. The labels are generated by hyperref package for the bookmarks. One can see them in file `foo.out` in Sphinx latex build repertory corresponding to file `foo.tex`.

The titles themselves are not very human-friendly in file `foo.out`: in UTF-16BE format, PDF-escaped ([direct reference to relevant section in hyperref pdf doc](http://mirrors.ctan.org/macros/latex/contrib/hyperref/doc/manual.pdf#nameddest=subsection.5.18)) but this is a detail.

- advantage of such preexisting targets is that the hypertarget opens the PDF right above the section title not immediately after it,

- disadvantage is that the actual names of destinations are the ones generated by hyperref itself although they seem quite guessable as the example above which is a link using ``manual.pdf#nameddest=subsection.5.18`` shows.

- of course a user chosen target name will remain identical throughout evolution of the document.

When the `bookmark` package  is loaded after `hyperref` the used file has extension `out.ps`, and its format is a bit different but for sectioning titles it seems the same.

My remarks above only explain that for sectioning units, the functionality is already existing with no change at Sphinx side... 

... @jessetan idea about enabling this for user-labeled captions is a bit different, +1 at any rate.

**Comment by jessetan** [comment]
Yes, there is already a label for section headers, but as you point out in your last bullet, a user-defined label will remain constant. An autogenerated label like "section.1.2" will no longer point to the intended part if sections are added. So for humans, autogenerated names with numbers are not really usable.

**Comment by jfbu** [comment]
I document why ``\hypertarget`` can not serve as general solution (hence the more complicated latex approach in #8808).

LaTeX test file (remove txt extension)
[testhypertarget.tex.txt](https://github.com/sphinx-doc/sphinx/files/5910677/testhypertarget.tex.txt)

And pdf output 
[testhypertarget.pdf](https://github.com/sphinx-doc/sphinx/files/5910678/testhypertarget.pdf)
